From efc78de83e95a03cc0d45d1d61c4b338c60d7311 Mon Sep 17 00:00:00 2001
From: Michael Rogenmoser <michael@rogenmoser.us>
Date: Mon, 17 Jul 2023 16:51:58 +0200
Subject: [PATCH] regtool: Patch windowing bug

---
 reg_top.sv.tpl | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/reg_top.sv.tpl b/reg_top.sv.tpl
index 974388759..1834a8253 100644
--- a/reg_top.sv.tpl
+++ b/reg_top.sv.tpl
@@ -355,7 +355,7 @@ $ endif
   assign reg_intf_req = reg_intf_demux_req[${num_wins}];
   assign reg_intf_demux_rsp[${num_wins}] = reg_intf_rsp;
 
-    % for i,t in enumerate(block.wins):
+    % for i in range(num_wins):
   assign reg_req_win_o[${i}] = reg_intf_demux_req[${i}];
   assign reg_intf_demux_rsp[${i}] = reg_rsp_win_i[${i}];
     % endfor
-- 
2.31.1

