From 2949fcbc707604e28c35b03c6993d82064270dc8 Mon Sep 17 00:00:00 2001
From: Michael Rogenmoser <michael@rogenmoser.us>
Date: Mon, 17 Jul 2023 14:58:46 +0200
Subject: [PATCH] [PATCH] Fix issue with Solderpad license check

---
 regtool.py | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/regtool.py b/regtool.py
index ccce88f66..b3f8959b2 100755
--- a/regtool.py
+++ b/regtool.py
@@ -292,7 +292,10 @@ def main():
         if found_lunder:
             src_lic = found_lunder
         if found_spdx:
-            src_lic += '\n' + found_spdx
+            if src_lic is None:
+                src_lic = '\n' + found_spdx
+            else:
+                src_lic += '\n' + found_spdx
 
         with outfile:
             if fmt == 'registers':
-- 
2.31.1

